<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc" style="margin-top: 1em;"><ul class="toc-item"><li><span><a href="#関数" data-toc-modified-id="関数-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>関数</a></span></li><li><span><a href="#ディレクトリ設定" data-toc-modified-id="ディレクトリ設定-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>ディレクトリ設定</a></span></li><li><span><a href="#ファイルリストの差分リストを作成する" data-toc-modified-id="ファイルリストの差分リストを作成する-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>ファイルリストの差分リストを作成する</a></span></li><li><span><a href="#重複ファイルを探す" data-toc-modified-id="重複ファイルを探す-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>重複ファイルを探す</a></span><ul class="toc-item"><li><span><a href="#バックアップされていたファイルの数" data-toc-modified-id="バックアップされていたファイルの数-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>バックアップされていたファイルの数</a></span></li></ul></li><li><span><a href="#指定されたリストのファイルを削除するバッチファイルを出力する" data-toc-modified-id="指定されたリストのファイルを削除するバッチファイルを出力する-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>指定されたリストのファイルを削除するバッチファイルを出力する</a></span></li><li><span><a href="#リネーム" data-toc-modified-id="リネーム-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>リネーム</a></span></li><li><span><a href="#cp932チェック" data-toc-modified-id="cp932チェック-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>cp932チェック</a></span></li><li><span><a href="#バグリカバリー" data-toc-modified-id="バグリカバリー-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>バグリカバリー</a></span></li><li><span><a href="#隠しファイルのリネーム" data-toc-modified-id="隠しファイルのリネーム-9"><span class="toc-item-num">9&nbsp;&nbsp;</span>隠しファイルのリネーム</a></span><ul class="toc-item"><li><span><a href="#リストの作成" data-toc-modified-id="リストの作成-9.1"><span class="toc-item-num">9.1&nbsp;&nbsp;</span>リストの作成</a></span></li><li><span><a href="#いよいよリネーム" data-toc-modified-id="いよいよリネーム-9.2"><span class="toc-item-num">9.2&nbsp;&nbsp;</span>いよいよリネーム</a></span></li></ul></li></ul></div>

In [5]:
import pprint
%pwd

'C:\\Users\\papa\\Documents\\Git\\Ham\\hello-world'

# 関数

In [172]:
def grep2( key, lst ):
    for idx,s in enumerate(lst):
        for t in s:
            if key in t:
                print( idx, s)

In [173]:
def grep( key, lst ):
    for idx,s in enumerate(lst):
        if key in s:
            print( idx, s)

In [6]:
def print_list( lst ):
    for idx, item in enumerate(lst):
        print( idx,item )

In [7]:
def get_key( pasuname, idx=0 ):
    """
    パス名からバックアップシステムが付けた日付識別文字列を除去する
    """
    key_spo = " UTC)"
    dist = 21 # ' (201'
    pos_spo = pasuname.rfind( key_spo )
    pos_atm = pos_spo - dist
    if pasuname[pos_atm:pos_atm+5] != ' (201':
        print( idx, ":", "head != (201", pasuname, ' but ', pasuname[pos_atm:pos_atm+5] )
        print( "posｖ_atm= ", pos_atm, "pos_spo= ", pos_spo)
        assert(False)
    if ( pos_atm <= 0 ):
        print( idx, ":", "pos_atm <= 0", pos_atm)
        assert(False)
    return pasuname[0:pos_atm] + pasuname[pos_spo+5:]
    

In [8]:
def upath2dospath( list_slash ):
    list_back = []
    for st in list_slash:
        tmp_st = st.replace( '/','\\')
        tmp_st = '"' + tmp_st + '"'
        list_back.append( tmp_st )              
    return list_back
    

In [9]:
def fileout_list( fname, lst, fmt ):
    with open(fname,"wt",encoding='cp932') as f:
        for st in lst:
            f.write( fmt % (st) )
        f.close()   

In [54]:
def fileout_list2( fname, lst1, lst2, fmt ):
    with open(fname,"wt",encoding='cp932') as f:
        for st1, st2 in zip(lst1, lst2):
            f.write( fmt % (st1,st2) )
        f.close() 

In [116]:
def fileout_format_list( fname, lst, fmt ):
    with open(fname,"wt",encoding='cp932') as f:
        f.write( "@echo off\n\n")
        for st in lst:
            f.write( fmt.format(st) )
        f.write( "@echo on\n")
        f.write( "echo Done!!\n")
        f.close() 

In [233]:
def fileout_format_list2( fname, lst1, lst2, fmt ):
    with open(fname,"wt",encoding='cp932') as f:
        for st1, st2 in zip(lst1, lst2):
            f.write( fmt.format(st1,st2) )
        f.close()

In [131]:
def chk_cp932( lst ):
    for idx,s in enumerate(lst):
        try:
            s_enc = s.encode('cp932')
        except:
            print( idx, s)
        

# ディレクトリ設定

In [32]:
bpath = "Z:\\BackUp\\WD3T_20191110\\"
wpath = "Z:\\Documents and Settings\\"

# ファイルリストの差分リストを作成する

In [76]:
file1 = bpath + "file_damaged_D.txt" # utf-8 lf
file2 = bpath + "file_backup.txt"
out12 = bpath + "diff_damaged_D-backup.txt"
out21 = bpath + "diff_backup-damaged_D.txt"

In [4]:
with open(file1,encoding='utf-8') as f:
    file1_list = f.read().splitlines()
    f.close()
    
with open(file2,encoding='utf-8') as f:
    file2_list = f.read().splitlines()
    f.close()


In [5]:
file2_list = [ lst.replace('/Documents and Settings',"") for lst in file2_list ]
file1_set = set(file1_list)
file2_set = set(file2_list)
file12_set = sorted(list(file1_set - file2_set))
file21_set = sorted(list(file2_set - file1_set))

In [76]:
with open(out12,"wt",encoding='utf-8') as f:
    f.write("\n".join(file12_set))
    f.close()
with open(out21,"wt",encoding='utf-8') as f:
    f.write("\n".join(file21_set))
    f.close()

In [23]:
#%qtconsole

# 重複ファイルを探す

## バックアップされていたファイルの数
Microsoft Office の一時ファイル（~$で始まるファイル）がなぜか見えない。ひとまず除外して考える。<br>
* 更新ファイルを含めた総数 61919 ( 内一時ファイル 6 )
* 更新ファイルを除外した数 61709( 内一時ファイル 4 )<br>
* 復元するファイル数 61705


In [186]:
# file list を読み込んで file_list を作る
filename = "Z:\\work\\backup_file_list.txt" # utf-8
with open(filename,encoding='utf-8') as f:
    file_list = f.read().splitlines()
    f.close()

file_list.sort() # 重要！ 識別子の辞書順が大きいファイルが最新
print( len(file_list))

61919


In [41]:
# file list から~$ で始まるファイル（officeの一時ファイル）を除く
ari_file = [ lst for lst in file_list if '~$' in lst ]
nashi_file = [ lst for lst in file_list if '~$' not in lst ]
print( '%d file 中 ~$ を含むファイルは%d個。含まないファイルは%d個です。\n含むファイルは以下のとおりです。' %( len(file_list), len(ari_file), len(nashi_file) ) )
print_list( ari_file )

61919 file 中 ~$ を含むファイルは6個。含まないファイルは61913個です。
含むファイルは以下のとおりです。
0 ./Owner/My Documents/中学受験関係/~$日程 (2019_01_12 07_48_49 UTC).pptx
1 ./Owner/My Documents/中学受験関係/パパ作成資料/浮力の問題/~$問１ (2019_02_09 07_50_59 UTC).docx
2 ./Owner/My Documents/中学受験関係/過去問/中大横浜/算数/~$まとめ (2019_01_05 16_42_51 UTC).pptx
3 ./Owner/My Documents/中学受験関係/過去問/中大横浜/算数/~$まとめ (2019_01_06 08_08_29 UTC).pptx
4 ./Owner/My Documents/中学受験関係/過去問/洗足/~$過去問傾向 (2019_01_05 11_39_24 UTC).xlsx
5 ./Owner/My Documents/中学受験関係/過去問/洗足/~$過去問傾向 (2019_02_03 16_37_15 UTC).xlsx


In [42]:
file_list = nashi_file
file_list.sort()

In [217]:
# file_list から更新ファイルを一つのリストにまとめて、そのリストをres_list とする    
res_list = []
cur_list = [file_list[0]]
cur_key = get_key( file_list[0] )

for idx, lst in enumerate( file_list[1:] ):
    tmp_key = get_key( lst, idx)
    if cur_key == tmp_key:
        cur_list.append( lst )
    else:
        res_list.append( cur_list )
        cur_list = [ lst ]
        cur_key = tmp_key        
res_list.append( cur_list) # 最終エントリーをフラッシュ
len(res_list)    

61709

In [44]:
# 重複リストを作成
dup_list = []
for lst in res_list:
    if len(lst) > 1 :
        dup_list.append(lst)
        for l in lst:
            print( l )
        print('\n')

./All Users/Documents/My Pictures/Camera/IXY3/2017/155___06/MVI_5828 (2019_01_05 05_46_06 UTC).MOV
./All Users/Documents/My Pictures/Camera/IXY3/2017/155___06/MVI_5828 (2019_10_27 04_13_21 UTC).MOV


./All Users/Documents/My Pictures/Camera/IXY3/2019/173___03/選抜/IMG_7104 (2019_04_14 03_15_05 UTC).JPG
./All Users/Documents/My Pictures/Camera/IXY3/2019/173___03/選抜/IMG_7104 (2019_04_14 05_38_19 UTC).JPG


./All Users/Documents/My Pictures/Camera/KissDX/2019/0101/選抜/IMG_8712 (2019_04_13 00_29_04 UTC).CR2
./All Users/Documents/My Pictures/Camera/KissDX/2019/0101/選抜/IMG_8712 (2019_04_13 05_50_13 UTC).CR2


./All Users/Documents/My Pictures/Camera/KissDX/2019/0101/選抜/IMG_8726 (2019_04_13 00_29_04 UTC).CR2
./All Users/Documents/My Pictures/Camera/KissDX/2019/0101/選抜/IMG_8726 (2019_04_13 05_50_13 UTC).CR2


./All Users/Documents/My Pictures/Camera/KissDX/2019/0101/選抜/IMG_8737 (2019_04_13 00_29_04 UTC).CR2
./All Users/Documents/My Pictures/Camera/KissDX/2019/0101/選抜/IMG_8737 (2019_04_13 05_50_13

In [45]:
# 最新リストを作成
latest_list = []
for lst in res_list:
    latest_list.append(lst[-1])
print( len(latest_list) )


61705


In [46]:
# 最新を除く重複リストを作成
notlatest_list = []
for lst in res_list:
    if len(lst) > 1:
        notlatest_list.append(lst[:-1])
print_list( notlatest_list )

0 ['./All Users/Documents/My Pictures/Camera/IXY3/2017/155___06/MVI_5828 (2019_01_05 05_46_06 UTC).MOV']
1 ['./All Users/Documents/My Pictures/Camera/IXY3/2019/173___03/選抜/IMG_7104 (2019_04_14 03_15_05 UTC).JPG']
2 ['./All Users/Documents/My Pictures/Camera/KissDX/2019/0101/選抜/IMG_8712 (2019_04_13 00_29_04 UTC).CR2']
3 ['./All Users/Documents/My Pictures/Camera/KissDX/2019/0101/選抜/IMG_8726 (2019_04_13 00_29_04 UTC).CR2']
4 ['./All Users/Documents/My Pictures/Camera/KissDX/2019/0101/選抜/IMG_8737 (2019_04_13 00_29_04 UTC).CR2']
5 ['./All Users/Documents/My Pictures/Camera/KissDX/2019/0101/選抜/IMG_8769 (2019_04_13 00_29_04 UTC).xmp']
6 ['./All Users/Documents/My Pictures/Camera/KissDX/2019/0101/選抜/IMG_8773 (2019_04_13 00_29_04 UTC).CR2']
7 ['./All Users/Documents/My Pictures/Camera/KissDX/2019/0501/選抜/IMG_8873 (2019_07_20 03_00_38 UTC).psd', './All Users/Documents/My Pictures/Camera/KissDX/2019/0501/選抜/IMG_8873 (2019_07_20 10_35_07 UTC).psd']
8 ['./All Users/Documents/My Pictures/Camera/Kis

# 指定されたリストのファイルを削除するバッチファイルを出力する

In [134]:
notlatest_list_f = []
for lst in notlatest_list:
    notlatest_list_f = notlatest_list_f + lst
notlatest_list_f_dos = upath2dospath( notlatest_list_f )
delbat_filename  = wpath + "delbat.bat"

In [50]:
fileout_list( delbat_filename, notlatest_list_f_dos, "del /f %s\n")

<pre>    
papa@shiro:/mnt/z/work$ wc backup_file_deldup_list.txt
  61711  372047 6463089 backup_file_deldup_list.txt
papa@shiro:/mnt/z/work$ wc backup_file_list.txt
  61919  373277 6483214 backup_file_list.txt
</pre>  

 208 ファイル消去（ del /f filename で”ファイルが見つかりませんでした”が3個でたがちゃんと消えている。不思議）

# リネーム
* 隠しファイルはMSDOSのRENではリネームできないのでchkバッチ走らせて検出されたら手で修正と思ったが面倒。最初の１つだけリネームしてしまったので、あとは放置することにした。*.info, *.GMD, Thumbs.db, ZbThumbnail.info, zb4meta.info

* 59664 ".\Owner\My Documents\ダンス用音楽編集\BurnItUp（ばーんいっとあっぷ）\Janet Jackson – BURNITUP! Feat. Missy Elliott (Audio Stream) (2019_01_05 05_46_06 UTC).mp4" はcp932 変換で失敗するので、手で修正

* リストのソートバグのため50000-59664 でファイル名の重複が発生。画面上では26個だが、chkスクリプトで数えると28個ある。被害のあったフォルダを元に戻して正しい処理を実施した方がよさそう

* 同じ原因で、59665-61705 までもファイル名の重複が１つだけ発生。元ファイルは次のような感じ
<pre>
    59954 ./Owner/My Documents/中学受験関係/チェックシート作成/共通問題 (2019_01_05 05_46_06 UTC).psd
    59955 ./Owner/My Documents/中学受験関係/チェックシート作成/共通問題 (2019_01_05 05_46_06 UTC).xlsx
    59956 ./Owner/My Documents/中学受験関係/チェックシート作成/共通問題 (2019_02_09 08_53_06 UTC).psd
</pre>
* 同上範囲で ".\Owner\My Documents\音\C2\rec_51.wav" が隠しファイルでリネーム失敗していので、通常ファイルに変更して、手動でリネーム

* とりあえず、一通りリネームを完了した。ファイル名の重複と隠しファイルの問題を次で考える。


In [124]:
len(latest_list) # officeの一時ファイル（６個）を含まない

61705

In [125]:
newname_list = []
newname_full_list = []
for str in latest_list:
    name = get_key( str )
    newname_list.append(name.split('/')[-1]) # DOS のren コマンドは変更後の名称にパスを含んではいけない
    newname_full_list.append( name )

In [126]:
latest_list_dos = upath2dospath( latest_list)
newname_list_dos = upath2dospath( newname_list)
newname_full_list_dos = upath2dospath( newname_full_list)

In [167]:
st = 59665
ed = len(latest_list_dos)
ren_filename = wpath + 'ren%d_%d.bat' % (st,ed )
chk_filename = wpath + 'chk%d_%d.bat' % (st,ed )
fileout_list2( ren_filename, latest_list_dos[st:ed], newname_list_dos[st:ed], "ren %s %s\n")
fileout_format_list( chk_filename, newname_full_list_dos[st:ed], "if not exist {0} echo {0}\n" )

# cp932チェック

In [132]:
chk_cp932( latest_list_dos)

59664 ".\Owner\My Documents\ダンス用音楽編集\BurnItUp（ばーんいっとあっぷ）\Janet Jackson – BURNITUP! Feat. Missy Elliott (Audio Stream) (2019_01_05 05_46_06 UTC).mp4"


# バグリカバリー
被害ファイルはchk50000-59664.bat の重複から推定すると
<pre>
".\All Users\Documents\My Pictures\Camera\KissDX\2019\0101\選抜\IMG_8769.CR2"
".\All Users\Documents\My Pictures\Camera\PEN\2019\0301\P3310017.JPG"
".\All Users\Documents\My Pictures\Camera\PEN\2019\0301\P3310017.ORF"
".\All Users\Documents\My Pictures\Camera\PEN\2019\0301\P4200070.JPG"
".\All Users\Documents\My Pictures\Camera\PEN\2019\0301\P4200070.ORF"
".\All Users\Documents\My Pictures\Camera\PEN\2019\0301\P4200071.JPG"
".\All Users\Documents\My Pictures\Camera\PEN\2019\0301\P4200071.ORF"
".\All Users\Documents\My Pictures\Camera\PEN\2019\0301\P4200072.JPG"
".\All Users\Documents\My Pictures\Camera\PEN\2019\0301\P4200072.ORF"
".\All Users\Documents\My Pictures\Camera\PEN\2019\0301\P4200073.JPG"
".\All Users\Documents\My Pictures\Camera\PEN\2019\0301\P4200073.ORF"
".\All Users\Documents\My Pictures\Camera\PEN\2019\0301\P4200074.JPG"
".\All Users\Documents\My Pictures\Camera\PEN\2019\0301\P4200074.ORF"
".\All Users\Documents\My Pictures\Camera\PEN\2019\0301\P4200075.JPG"
".\All Users\Documents\My Pictures\Camera\PEN\2019\0301\P4200075.ORF"
".\All Users\Documents\My Pictures\Camera\PEN\2019\0301\P4200076.JPG"
".\All Users\Documents\My Pictures\Camera\PEN\2019\0301\P4200076.ORF"
".\All Users\Documents\My Pictures\Camera\PEN\2019\0301\P4200077.JPG"
".\All Users\Documents\My Pictures\Camera\PEN\2019\0301\P4200077.ORF"
".\All Users\Documents\My Pictures\Camera\PEN\2019\0301\P4200078.JPG"
".\All Users\Documents\My Pictures\Camera\PEN\2019\0301\P4200078.ORF"
".\All Users\Documents\My Pictures\Camera\PEN\2019\0501_jpeg2\P5260034.jpg"
".\All Users\Documents\My Pictures\Camera\PEN\2019\0501_jpeg2\P5260035.jpg"
".\All Users\Documents\My Pictures\Camera\PEN\2019\0501_jpeg2\P5260037.jpg"
".\All Users\Documents\My Pictures\Camera\PEN\2019\0501_jpeg2\P5260077.jpg"
".\All Users\Documents\My Pictures\Camera\PEN\2019\0501_jpeg2\P5260113.jpg"
".\Owner\My Documents\My Music\さまざまなアーティスト\desktop.ini"
".\Owner\My Documents\My Music\MIDI\SONG\シアワセミマン.MID"
</pre>
最後の２つは uniq -d が以下のように漢字を含むケースに非対応のためっぽい。なので被害ファイルは 26 個でログ画面での数と一致。
<pre>
".\Owner\My Documents\My Music\さまざまなアーティスト\desktop.ini"
".\Owner\My Documents\My Music\ケルティック・ウーマン\desktop.ini"
</pre>

移行先でリネームしそびれて残っているファイルが最新のはずなので、手動で古い方を消してこれをリネームして完了。

# 隠しファイルのリネーム
_xcopy src dst /h /k_ を使う<br>
消すのは、アトリビュートに隠しファイルとリードオンリーを指定して_del /ahr src_ <br>
２つのコマンドでオプションの位置が違うが仕様らしい。<br>
ファイルは全部リードオンリーになっているみたい。

## リストの作成

In [185]:
# file list を読み込んで file_list を作る
filename = "Z:\\work\\kakushi_file.txt" # utf-8
with open(filename,encoding='utf-8') as f:
    kks_file_list = f.read().splitlines()
    f.close()

kks_file_list.sort()

In [218]:
# kks_file_list から更新ファイルを一つのリストにまとめて、そのリストをkks_res_list とする    
kks_res_list = []
cur_list = [kks_file_list[0]]
cur_key = get_key( kks_file_list[0] )

for idx, lst in enumerate( kks_file_list[1:] ):
    tmp_key = get_key( lst, idx)
    if cur_key == tmp_key:
        cur_list.append( lst )
    else:
        kks_res_list.append( cur_list )
        cur_list = [ lst ]
        cur_key = tmp_key        
kks_res_list.append( cur_list) # 最終エントリーをフラッシュ
len(kks_res_list)

337

In [222]:
# 最新を除く重複リストを作成( MSDOSで消せないのでWSLでrmする)
kks_notlatest_list = []
for lst in kks_res_list:
    if len(lst) > 1:
        kks_notlatest_list.append(lst[:-1])
print_list( kks_notlatest_list )

0 ['./Owner/My Documents/中学受験関係/過去問/中大横浜/算数/~$まとめ (2019_01_05 16_42_51 UTC).pptx']
1 ['./Owner/My Documents/中学受験関係/過去問/洗足/~$過去問傾向 (2019_01_05 11_39_24 UTC).xlsx']


In [223]:
# 最新リストを作成
kks_latest_list = []
for lst in kks_res_list:
    kks_latest_list.append(lst[-1])
print( len(kks_latest_list) )


337


## いよいよリネーム

In [228]:
kks_newname_list = [ get_key(str) for str in kks_latest_list ]
kks_latest_list_dos = upath2dospath( kks_latest_list )
kks_newname_list_dos = upath2dospath( kks_newname_list )

In [253]:
st = 300
ed = len(kks_latest_list_dos)
ren_filename = wpath + 'kks_ren%d_%d.bat' % (st,ed )
chk_filename = wpath + 'kks_chk%d_%d.bat' % (st,ed )
fileout_format_list2( ren_filename, kks_latest_list_dos[st:ed], kks_newname_list_dos[st:ed], "xcopy   {0} {1} /h /k \nif exist {1} del /ahr {0}\n")
fileout_format_list( chk_filename, kks_newname_list_dos[st:ed], "if not exist {0} echo {0}\n" )

In [254]:
ed

337

完了